### PR TITLE
Use builtin font to draw powerline symbols

### PR DESCRIPTION
**Pull request by kchibisov**
In addition to box drawing it was decided to also draw powerline
symbols, since those are quite common and rather simple to draw with
present box drawing infra.

### CODE REVIEW

**Comment by kchibisov** [comment]
Updated to be against v0.11.0.

**Review by chrisduerr — CHANGES_REQUESTED** [review]
The glyphs don't appear to be perfect:

tmp

Especially with powerline symbols, the addition really only makes sense when things are absolutely perfect. If the slope doesn't start **immediately**, that's likely not something that's gonna be noticed, but if the slope starts too late like this or it's uneven that's gonna be really jarring (I was immediately able to notice this for example).

**Comment by solonovamax** [comment]
> The glyphs don't appear to be perfect:

I'm not sure how you arrived at this.

I've built & tested the branch, and from taking a screenshot and then counting pixels in GIMP, it seems to be accurate to me.

See:
image
image
image
image
image

**Comment by chrisduerr** [comment]
@solonovamax Your images show the same issue I've demonstrated in my screenshot. You just need to take a closer look.

**Comment by solonovamax** [comment]
Could you point out where it is specifically?
I believe that may help with resolving the issue, as if I can't notice it, OP might not be able to notice it. (Unless you've discussed this with them off-platform)

**Comment by chrisduerr** [comment]
> I believe that may help with resolving the issue, as if I can't notice it, OP might not be able to notice it. (Unless you've discussed this with them off-platform)

I believe he's capable of speaking for himself.

**Comment by solonovamax** [comment]
> > I believe that may help with resolving the issue, as if I can't notice it, OP might not be able to notice it. (Unless you've discussed this with them off-platform)
> 
> I believe he's capable of speaking for himself.

I don't want to imply that they can't speak for themselves, I just want to see this merged sooner as this is a feature I would like.

And I apologize if anything I said came off as ride; that was not my intent.

Though, I'll stop commenting unless I have fixes/etc, as this is now becomming off-topic for this thread.

**Comment by kchibisov** [comment]
Does latest commit changes anything @chrisduerr . I really don't see issues with filled rectangles.

picture

**Comment by chrisduerr** [comment]
Since you asked on IRC, this still doesn't look right to me:

tmp

Pretty sure it's because of the anti-aliasing, but there's a really sharp "edge" for the first pixel step while the following pixel steps look far softer, making it look wonky.

**Comment by Pytness** [comment]
Here is what I have came up to.

I use a Canvas twice the size of the glyph, draw the lines fill if necessary, and scale down the canvas to get some sort of antialiasing.
I have tried to line drawing algorithms, Xiaolin's (the one @kchibisov uses)  and Bresenham's.


Examples of other "rendering" methods (made with gimp)
image



Zoomed in:
image

**Comment by sharkusk** [comment]
Would it make sense to support a fallback/override font, where specific code point ranges can be set and used for characters in these ranges?

For example, a user could install a single Nerd Font as the override, then set the specific ranges to support whatever Gyphs they desire (https://github.com/ryanoasis/nerd-fonts/wiki/Glyph-Sets-and-Code-Points).  This would eliminate the need to use patched fonts or handle special symbol rendering in alacritty.

Config would look something like this:

```
font:
  # Special Symbol Override Font
  override:
    # Font family 
    family: Hasklug Nerd Font

   # Nerd Font Code Points
    ranges: 23fb-23fe, 2665, 26a1, 2b58, e000-e00a, e0a0-e0a2, e0a3, e0b0-e0b3, e0b4-e0c8, etc... 
```

**Comment by chrisduerr** [comment]
> Would it make sense to support a fallback/override font, where specific code point ranges can be set and used for characters in these ranges?

No, that would be complete nonsense. This patch is precisely because we **do not** want to use fonts for these glyphs. You can already setup fallback fonts through fontconfig just fine.

File: alacritty/src/renderer/text/builtin_font.rs [hunk 1]
```diff
@@ -495,6 +497,125 @@ fn box_drawing(character: char, metrics: &Metrics, offset: &Delta<i8>) -> Raster
...
+
+    let mut canvas = Canvas::new(width, height);
+
+    let x_end = width as f32 - 1.;
+    let y_end = height as f32 - 1.;
+    let y_center = y_end / 2.;
+
+    // Offset to achieve 90 degree angle in the tip when we can do so just by moving 1px down/up
+    // on the opposite side. When we can't do so, just cut on the tip side.
+    let y_offset = if (height + 1).saturating_sub(2 * width) / 2 == 1 { 1 } else { 0 };
```
Comment (chrisduerr) [inline alacritty/src/renderer/text/builtin_font.rs:514]: This is optional, it might be simpler and easier to maintain if we just always do this.

**Review by chrisduerr — CHANGES_REQUESTED** [review]

File: alacritty/src/renderer/text/builtin_font.rs [hunk 2]
```diff
@@ -495,6 +497,125 @@ fn box_drawing(character: char, metrics: &Metrics, offset: &Delta<i8>) -> Raster
...
+    let width = (metrics.average_advance as i32 + offset.x as i32) as usize;
+    // Use one eight of the cell width, since this is used as a step size for block elements.
+    let stroke_size = cmp::max((width as f32 / 8.).round() as usize, 1) as f32;
+
+    let mut canvas = Canvas::new(width, height);
+
+    let y_center = (height - 1) as f32 / 2.;
+    // Start with offset `1` and draw until the intersection of the 2 f(x) = x + 1 and
+    // g(x) = H - x + 1. The intersection happens at 2 * f(x) = H - 2 or f(x) = H/2,
+    // which is `y_center`.
```
Comment (chrisduerr) [inline alacritty/src/renderer/text/builtin_font.rs:511]: Nobody will ever understand what this comment means in a week or two lol.

**Review by chrisduerr — APPROVED** [review]

File: alacritty/src/renderer/text/builtin_font.rs (hunk 2 above)
Comment (kchibisov) [reply alacritty/src/renderer/text/builtin_font.rs]: I've tuned it a bit.

**Pull request merged**